Max nb columns in output reports : computing an accurate max number of columns (based on synthesis reports)

guilpier-code opened the pull request:
The current PR follows #1159, which is merely a preparatory cleaning PR.
The current PR adresses the real problem of issue #927 :  computing an accurate max number of columns an output report can contain.

Note that, as synthesis reports always contain more columns than any other report, the max number of columns computation is based on the number of columns an output variable can take in a synthesis report.

**Execution steps** : 
- Creating all output variables print info list, that is filling **variablesPrintInfo** where each variable's print status is **true**
  See **parameters.cpp** : `RetrieveVariableList(collector)`.
  At this point, clusters or BCs are not loaded, we can't give a  max column width to the dynamic or BCs output variables.
- Collecting print info status for each output variable (see reading functions in **parameters.cpp**)
- Collecting the max column width of each output variables when clusters or BCs are loaded.
  This is done inside **solver.hxx** : `variables.initializeFromStudy(study)`
  See the calls to `supplyMaxNumberOfColumns(study)`
- Computing the max number of columns a report can contains.
  This is done when calling **solver.hxx** : parameters.variablesPrintInfo.computeMaxColumnsCountInReports();
  By looping over all the kinds of generated reports, getting the number of columns they contain,  and storing the largest size. 
- Getting the already computed largest size in `SurveyResults::initializeMaxVariables`, just before printing

guilpier-code committed:
max nb columns for print : computing an accurate max number of columns (based on synthesis reports)

Note that synthesis reports contain more columns than year by year reports.
So the computed max number of columns is actually the max number of columns in a synthesis report

flomnes commented:
General remark: are you sure you want to use the **synthesis** case ? For year-by-year, it overestimates the number of columns by a factor of ~4 (min, max, exp, std).

guilpier-code commented:
> General remark: are you sure you want to use the **synthesis** case ? For year-by-year, it overestimates the number of columns by a factor of ~4 (min, max, exp, std).

What we just did here is computing accurately was what done before approximately.
We always have had one column size for all reports, whether it is a synthesis report or a year by year report.
If we want to adapt the column size to the report we currently print, we can think about it.

flomnes commented on src/solver/variable/bindConstraints.hxx:
```diff
@@ -122,6 +122,15 @@ void BindingConstraints<NextT>::initializeFromStudy(Data::Study& study)
+    if (pBCcount > 0)
+    {
+        NextType& bc = pBindConstraints[0];
+        bc.setBindConstraintsCount(pBCcount);
```
We only want "<" or ">" binding constraints here

guilpier-code replied:
Ok, see this PR's top comment

guilpier-code replied:
That's we have I think : **pBCcount** is the number of inequality BCs, isn't it ?
Maybe we should rename it into **inequalityBCcount_**

flomnes replied:
Yes, please do

flomnes commented on src/solver/variable/surveyresults/surveyresults.cpp:
```diff
@@ -509,39 +509,8 @@ void SurveyResults::initializeMaxVariables(const Data::Study& s)
 
     // Getting the any report's max number of columns
-    maxVariables = s.parameters.variablesPrintInfo.getMaxColumnsCount();
+    maxVariables = s.parameters.variablesPrintInfo.getTotalMaxColumnsCount();
```
NICE !

flomnes reviewed (commented):
Can you briefly explain the logic of your dev ? What steps are involved, who's responsible for what, etc.

guilpier-code commented:
> Can you briefly explain the logic of your dev ? What steps are involved, who's responsible for what, etc.

Ok, see this PR's top comment

guilpier-code commented on src/solver/variable/bindConstraints.hxx:
```diff
@@ -122,6 +122,15 @@ void BindingConstraints<NextT>::initializeFromStudy(Data::Study& study)
         bc.getPrintStatusFromStudy(study);
     }
+
+    // This is ugly (it's a work around). We should try to improve this.
```
As commented in the code says, we have a workaround here.
Instead of : 
```C++
if (pBCcount > 0)
{
     NextType& bc = pBindConstraints[0];
     bc.setBindConstraintsCount(pBCcount);
     bc.supplyMaxNumberOfColumns(study);
}
```
we should have : 
```C++
NextType::supplyMaxNumberOfColumns(study);
```
as it is done for **Areas** and **Links**.
This is due to the fact that BCs output variables are not **dynamic** output variables.
I left a comment on this **directly in the code** in the (following) PR #1180

flomnes replied:
I think this workaround is OK for now. At least we no longer have hard-coded constants such as "4 variables for each thermal clusters" and so on.

guilpier-code committed:
max nb columns for print : more cleaning (#1180)

* max nb columns for print : code improvement

Improvements :
- At reading the output variable print status (parameters.cpp)
- Small changes inside classes of variable-print-info.h/.cpp
- Removing function SurveyResults::initializeMaxVariables(...), which has shrunk until it does one function call.

* max nb columns for print : in class AllVariablesPrintInfo, changing the std::vector into a std::map

* max nb columns for print : renaming in output variable BC MARG COST

* max nb columns for print : fix a bug

* max nb columns for print : fix other bugs

* max nb columns for print : refresh names lists each time a print status is changed

* max nb columns for print : removing useless isEmpty()

Indeed, allVarsPrintInfo can never be empty, has it is build from code built at compile time

* max nb columns for print : try fix a compilation failure on Linux

* max nb columns for print : change for more readable manner to traverse std::map

* max nb columns for print : corrections after review

* max nb columns for print : more corrections after review

* max nb columns for print : amend previous commit (causes compilation error on Linux)

* max nb columns for print : remove lists that have to be updated in real-time

* Use enum value in template param (#1193)

* max nb columns for print : adding member function numberOfEnabledVariables() to class AllVariablesPrintInfo

---------

Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>
